### PR TITLE
New function pgr contraction hierarchies 3.8

### LINKED ISSUE #2536: Support for contraction hierarchies  (pgr_contractionHierarchies)
Hello the pgRouting team,

I recently worked with pgRouting on a graph that contained around 31 millions of edges and got nice results in terms of computation time (a few seconds). You're doing a really nice job for those who want to compute routes in a database context!

Anyway, in a lot of routing situations, we want to get results in less than one second, and I was wondering if some optimization could be possible for pgRouting.

## TL;DR

I read on the [GsoC 2022 program](https://github.com/pgRouting/pgrouting/wiki/GSoC-Ideas%3A-2022#details-of-idea-14) that implementing Contraction Hierarchies is still a valid idea. Hence comes my question: are you interested in a contribution on that topic? Do you still consider the Contraction Hierarchies algorithm has an added value for pgRouting?

## Contraction in pgRouting

Knowing that Contraction Hierarchies are a game changer for fast routing purpose, I was wondering if they were implemented in pgRouting. Then I read the [pgRouting documentation](https://docs.pgrouting.org/latest/en/contraction-family.html), digged into the [pgRouting mailing list archive](https://lists.osgeo.org/pipermail/pgrouting-dev/2015-December/001607.html) and read #440 (which confuses me a bit).

By cross-referencing information from the [GsoC 2016 contraction report](https://github.com/pgRouting/pgrouting/wiki/GSoC-2016-Contraction), I understand that the implemented features are dead end and linear contractions, and that Contraction Hierarchies are not implemented. Is it correct?

Another confusing point for me is that there are [CH vertices](https://github.com/pgRouting/pgrouting/blob/main/include/cpp_common/ch_vertex.h), [CH edges](https://github.com/pgRouting/pgrouting/blob/main/include/cpp_common/ch_edge.h) and [CH graphs](https://github.com/pgRouting/pgrouting/blob/main/include/contraction/ch_graphs.hpp) in the pgRouting API: in this context I hypothesize that "CH" means "Contraction Hierarchies", but I may be wrong?

## Improvement

As far as I understand the different contraction algorithms, there is a seminal difference between the pgRouting contractions and the Contraction Hierarchies. While the formers act on the graph topology, the latter considers the edge weights and evaluates every vertex in the graph in order to build a node ordering (the higher the node is in the hierarchy, the less likely its contraction is). Of course, it means that the contracted graph is valid as long as the weights are static.

In order to implement Contraction Hierarchies, one should first implement this node ordering procedure, whose outputs are a set of shortcuts (as in other contraction algorithms) and a rank in the hierarchy for each vertex (this is the big difference). This preprocessing could be another contraction referenced in [include/contraction/pgr_contract.hpp](https://github.com/pgRouting/pgrouting/blob/main/include/contraction/pgr_contract.hpp).

Then comes the routing query itself, that has to be run in the ordered graph (with shortcuts). This query algorithm is closed to a bidirectional Dijkstra, but highly depends on the vertex ordering. I do not know where could be the best place for such a querying algorithm; could a brand new `chquery` subfolder do the job?

## External resources

The Contraction Hierarchies is mainly described by the following papers:

- [Geisberger, 2008](http://algo2.iti.kit.edu/documents/routeplanning/geisberger_dipl.pdf)
- [Geisberger et al., 2008](http://algo2.iti.kit.edu/schultes/hwy/contract.pdf)
- [Geisberger et al., 2012](https://publikationen.bibliothek.kit.edu/1000028701)

As a side note, there is a [Github repo](https://github.com/RoutingKit/RoutingKit) handled by the research team, with (amongst other things) a [C++ implementation of Contraction Hierarchies](https://github.com/RoutingKit/RoutingKit/blob/master/src/contraction_hierarchy.cpp). Considering this could help to design a proof of concept within the pgRouting API.

### PR DESCRIPTION
**Pull request by cvvergara**
Fixes #2536  

Changes proposed in this pull request:

New function:
`pgr_contractionHierarchies(Edges SQL,  [directed, forbidden])`

Includes:
- documentation
- SQL and C/C++ implementation
- Basic unit pgtap tests
- Example executions for the documentation

Author: Aurélie Bousquet

@pgRouting/admins

## Summary by CodeRabbit

- **New Features**
  - Introduced the experimental function `pgr_contractionHierarchies` for building contraction hierarchies, enabling efficient shortest path queries by adding shortcut edges and hierarchical vertex ordering.
  - Added support for specifying forbidden vertices in contraction hierarchies computations.
  - Provided comprehensive documentation and usage examples for the new function.

- **Documentation**
  - Updated and expanded documentation to describe the new contraction hierarchies algorithm, including detailed explanations, diagrams, and SQL examples.
  - Added new documentation sections, images, and updated release notes and contributor lists.

- **Tests**
  - Added extensive test suites for contraction hierarchies, covering edge cases, type checks, and integration scenarios for both directed and undirected graphs.

- **Chores**
  - Included new sample data and updated build/test scripts to support contraction hierarchies testing and documentation.

### CODE REVIEW

File: tools/testers/contractionHierarchies_data.sql [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+3(0.5,1)
+4(1,0)
+5(1.5,0)
+6(3,0)
+7(3,1)
+8(2.5,1)
+9(2,2)
+10(1.5,1)
+11(1,2)
+*/
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply tools/testers/contractionHierarchies_data.sql:71]: Not doing in this PR

File: tools/testers/contractionHierarchies_data.sql [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    ( 2, 0, 0, st_point(0,0) ),
+    ( 3, 0.5, 1, st_point(0.5, 1) ),
+    ( 4, 1, 0, st_point(1, 0) ),
+    ( 5, 1.5, 0, st_point(1.5, 0) ),
+    ( 6, 3, 0, st_point(3, 0) ),
+    ( 7, 3, 1, st_point(3, 1) ),
+    ( 8, 2.5, 1, st_point(2.5, 1) ),
+    ( 9, 2, 2, st_point(2, 2) ),
+    ( 10, 1.5, 1, st_point(1.5, 1) ),
+    ( 11, 1, 2, st_point(1, 2) );
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply tools/testers/contractionHierarchies_data.sql:85]: Not needed in this small example

File: tools/testers/contractionHierarchies_data.sql [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+-- Sample graph #2
+DROP TABLE IF EXISTS edges_2;
+DROP TABLE IF EXISTS vertices_2;
+
+create table vertices_2 (
+    id integer primary key,
+    x double precision,
+    y double precision,
+    geom Geometry(Point)
+);
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply tools/testers/contractionHierarchies_data.sql:20]: Not needed in this small example

File: tools/testers/contractionHierarchies_data.sql [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+create table edges_2 (
+    id bigserial primary key,
+    source bigint,
+    target bigint,
+    cost float,
+    reverse_cost float,
+    geom geometry(Linestring)
+);
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply tools/testers/contractionHierarchies_data.sql:29]: Not needed in this small example

File: doc/src/release_notes.rst [hunk 5]
```diff
@@ -74,6 +74,8 @@ milestone for 3.8.0
     pgr_contractionDeadEnd new contraction function
   * `#2791 <https://github.com/pgRouting/pgrouting/issues/2791>`__:
     pgr_contractionLinear new contraction function
+  * `#2536 <https://github.com/pgRouting/pgrouting/issues/2536>`__:
+    Support for contraction hierarchies (pgr_contractionHierarchies)
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply doc/src/release_notes.rst:78]: The text comes from the title of issue which is not changing.

File: NEWS.md [hunk 6]
```diff
@@ -43,6 +43,8 @@ milestone for 3.8.0
     pgr_contractionDeadEnd new contraction function
   * [#2791](https://github.com/pgRouting/pgrouting/issues/2791):
     pgr_contractionLinear new contraction function
+  * [#2536](https://github.com/pgRouting/pgrouting/issues/2536):
+    Support for contraction hierarchies (pgr_contractionHierarchies)
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply NEWS.md:47]: The script was used to generate the contents of the file

File: include/drivers/contraction/contractionHierarchies_driver.h [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        contractionHierarchies_rt**,
+        size_t*,
+        char**,
+        char**,
+        char**);
+#ifdef __cplusplus
+}
+#endif
+
+#endif  // INCLUDE_DRIVERS_CONTRACTION_CONTRACTIONHIERARCHIES_DRIVER_H_
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply include/drivers/contraction/contractionHierarchies_driver.h:73]: This year is 2025 so the value is correct

File: pgtap/contraction/contractionHierarchies/edge_cases/undirected.pg [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    FROM (
+        VALUES
+            ('v'::CHAR, 1::BIGINT, ARRAY[]::BIGINT[], -1::BIGINT, -1::BIGINT, -1::FLOAT, -1::BIGINT, 1::BIGINT),
+            ('v', 11, ARRAY[]::BIGINT[], -1, -1, -1, 0, 2)
+    )
+    AS t(type, id, contracted_vertices, source, target, cost, metric, vertex_order);
+
+    RETURN QUERY
+    SELECT set_eq('q1', 'sol1');
+
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply pgtap/contraction/contractionHierarchies/edge_cases/undirected.pg:65]: The file is not to be run interactively, so discarding the prepared statements on the `ROLLBACK` is ok.

File: pgtap/contraction/contractionHierarchies/edge_cases/undirected.pg [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+ ********************************************************************PGR-GNU*/
+
+BEGIN;
+
+UPDATE edges
+SET
+    cost = sign(cost),
+    reverse_cost = sign(reverse_cost);
+
+SELECT CASE WHEN NOT min_version('3.8.0') THEN plan(1) ELSE plan(6) END;
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply pgtap/contraction/contractionHierarchies/edge_cases/undirected.pg:31]: We will remember

File: src/contraction/contractionHierarchies_driver.cpp [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+void
+pgr_contractionHierarchies(
+        char *edges_sql,
+        ArrayType* forbidden,
+        bool directed,
+        contractionHierarchies_rt **return_tuples,
+        size_t *return_count,
+        char **log_msg,
+        char **notice_msg,
+        char **err_msg) {
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply src/contraction/contractionHierarchies_driver.cpp:56]: The caller is within the scope of the project, and its passing a reference to a variable.

File: docqueries/contraction/contractionHierarchies.result [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+ v    | 13 | {}                  |     -1 |     -1 |   -1 |     -1 |           11
+ v    | 14 | {}                  |     -1 |     -1 |   -1 |      0 |           16
+ v    | 15 | {}                  |     -1 |     -1 |   -1 |     -1 |           13
+ v    | 16 | {}                  |     -1 |     -1 |   -1 |     -1 |           12
+ v    | 17 | {}                  |     -1 |     -1 |   -1 |      0 |           17
+ e    | -1 | {7}                 |     11 |      8 |    2 |     -1 |           -1
+ e    | -2 | {7,8}               |     11 |      9 |    3 |     -1 |           -1
+ e    | -3 | {8}                 |     12 |      9 |    2 |     -1 |           -1
+ e    | -4 | {11}                |     12 |     16 |    2 |     -1 |           -1
+(21 rows)
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply docqueries/contraction/contractionHierarchies.result:56]: The docqueries files, show the output as is, it is not for testing purposes, it is for documentation example use.

**Review by robe2 — APPROVED** [review]

**Comment by cvvergara** [comment]
The function is added as experimental as it still has a lot of issues, some are noticed by the rabbit, others are for code organization. 
When the issues are fixed then we can promote to proposed.

**Comment by cvvergara** [comment]
@coderabbitai create an issue with all the unresolved conversations
